File: backend/app/schemas/decision.py
"""Pydantic schemas for Decision API."""

from datetime import datetime
from enum import Enum
from typing import Optional, Any

from pydantic import BaseModel, Field, validator


class DecisionCategory(str, Enum):
    """Decision category types."""
    STRATEGIC = "strategic"
    OPERATIONAL = "operational"
    TACTICAL = "tactical"
    FINANCIAL = "financial"
    TECHNICAL = "technical"


class DecisionStatus(str, Enum):
    """Decision status values."""
    ACTIVE = "active"
    ARCHIVED = "archived"
    IMPLEMENTED = "implemented"
    ABANDONED = "abandoned"


class FactorCategory(str, Enum):
    """Factor category types."""
    MARKET = "market"
    FINANCIAL = "financial"
    TECHNICAL = "technical"
    ORGANIZATIONAL = "organizational"
    CUSTOMER = "customer"
    COMPETITIVE = "competitive"
    REGULATORY = "regulatory"
    STRATEGIC = "strategic"


class ImpactLevel(str, Enum):
    """Impact level values."""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class Timeframe(str, Enum):
    """Timeframe values."""
    SHORT_TERM = "short-term"
    MEDIUM_TERM = "medium-term"
    LONG_TERM = "long-term"


# Request Schemas

class DecisionCreate(BaseModel):
    """Schema for creating a new decision."""
    title: str = Field(..., min_length=3, max_length=255)
    description: str = Field(..., min_length=10)
    category: Optional[DecisionCategory] = None
    decision_date: Optional[datetime] = None
    context: Optional[dict[str, Any]] = None

    @validator('title')
    def title_not_empty(cls, v):
        if not v.strip():
            raise ValueError('Title cannot be empty')
        return v.strip()


class DecisionUpdate(BaseModel):
    """Schema for updating a decision."""
    title: Optional[str] = Field(None, min_length=3, max_length=255)
    description: Optional[str] = Field(None, min_length=10)
    category: Optional[DecisionCategory] = None
    status: Optional[DecisionStatus] = None
    decision_date: Optional[datetime] = None
    context: Optional[dict[str, Any]] = None


class DecisionQuery(BaseModel):
    """Schema for querying past decisions."""
    query: str = Field(..., min_length=3, max_length=500)
    include_timeline: bool = Field(default=True)
    include_factors: bool = Field(default=True)
    max_results: int = Field(default=10, ge=1, le=50)


# Response Schemas

class FactorResponse(BaseModel):
    """Schema for decision factor response."""
    id: str
    decision_id: str
    name: str
    category: FactorCategory
    impact_score: int = Field(..., ge=1, le=10)
    explanation: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class OutcomeResponse(BaseModel):
    """Schema for decision outcome response."""
    id: str
    decision_id: str
    description: str
    outcome_type: str
    likelihood: Optional[int] = Field(None, ge=0, le=100)
    impact: Optional[ImpactLevel] = None
    timeframe: Optional[Timeframe] = None
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DecisionResponse(BaseModel):
    """Schema for decision response."""
    id: str
    user_id: str
    title: str
    description: str
    category: Optional[str] = None
    status: str
    context: Optional[dict[str, Any]] = None
    graph_node_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    decision_date: Optional[datetime] = None
    factors: list[FactorResponse] = []
    outcomes: list[OutcomeResponse] = []

    class Config:
        from_attributes = True


class TimelineEvent(BaseModel):
    """Schema for timeline event."""
    date: str
    type: str
    title: str
    is_main: bool = False
    relationship: Optional[str] = None


class TimelineResponse(BaseModel):
    """Schema for decision timeline response."""
    decision_id: str
    decision_title: str
    events: list[TimelineEvent]


class RelatedDecision(BaseModel):
    """Schema for related decision."""
    id: str
    title: str
    date: str
    relationships: list[str]
    distance: int


class RelatedDecisionsResponse(BaseModel):
    """Schema for related decisions response."""
    decision_id: str
    related: list[RelatedDecision]


class GraphStats(BaseModel):
    """Schema for graph statistics."""
    decisions: int
    people: int
    projects: int
    factors: int
    relationships: int


class EntityExtraction(BaseModel):
    """Schema for extracted entities."""
    decisions: list[str]
    people: list[str]
    projects: list[str]
    dates: list[str]
    keywords: list[str]


class FactorAnalysis(BaseModel):
    """Schema for factor analysis."""
    name: str
    category: FactorCategory
    impact_score: int = Field(..., ge=1, le=10)
    explanation: str


class FactorAnalysisResponse(BaseModel):
    """Schema for factor analysis response."""
    decision_id: str
    factors: list[FactorAnalysis]


class DecisionComparison(BaseModel):
    """Schema for decision comparison."""
    decision1_id: str
    decision2_id: str
    similarities: list[str]
    differences: list[str]
    relationship: str
    insights: list[str]


class PredictedOutcome(BaseModel):
    """Schema for predicted outcome."""
    description: str
    likelihood: int = Field(..., ge=0, le=100)
    impact: ImpactLevel
    timeframe: Timeframe


class OutcomePrediction(BaseModel):
    """Schema for outcome prediction."""
    decision_id: str
    outcomes: list[PredictedOutcome]
    risks: list[str]
    opportunities: list[str]


class DecisionList(BaseModel):
    """Schema for paginated decision list."""
    decisions: list[DecisionResponse]
    total: int
    page: int
    page_size: int


# Analysis Schemas

class AnalysisRequest(BaseModel):
    """Schema for decision analysis request."""
    decision_id: str
    include_timeline: bool = True
    include_factors: bool = True
    include_predictions: bool = True


class AnalysisResponse(BaseModel):
    """Schema for complete decision analysis."""
    decision: DecisionResponse
    timeline: Optional[TimelineResponse] = None
    factors: Optional[list[FactorAnalysis]] = None
    predictions: Optional[OutcomePrediction] = None
    related_decisions: Optional[list[RelatedDecision]] = None


# Phase 6: Pattern Analysis Schemas

class PatternTimespan(BaseModel):
    """Schema for pattern timespan."""
    start: str
    end: str


class DecisionPattern(BaseModel):
    """Schema for a detected decision pattern."""
    pattern_type: str
    description: str
    frequency: int
    decisions: list[str]
    timespan: PatternTimespan


class PatternAnalysisRequest(BaseModel):
    """Schema for pattern analysis request."""
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class PatternAnalysisResponse(BaseModel):
    """Schema for pattern analysis response."""
    patterns: list[DecisionPattern]
    insights: list[str]
    recommendations: list[str]


# Phase 6: AI Insights Schemas

class InsightType(str, Enum):
    """AI insight types."""
    OBSERVATION = "observation"
    RECOMMENDATION = "recommendation"
    RISK = "risk"
    OPPORTUNITY = "opportunity"


class AIInsight(BaseModel):
    """Schema for a single AI insight."""
    type: InsightType
    title: str
    description: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    related_factors: Optional[list[str]] = None


class AIInsightsResponse(BaseModel):
    """Schema for AI insights response."""
    decision_id: str
    insights: list[AIInsight]
    summary: str
    generated_at: datetime

File: frontend/types/decision.ts
/**
 * Decision Type Definitions for InnoSynth.ai Decision Archaeology
 */

// Enums matching backend
export type DecisionCategory = 'strategic' | 'operational' | 'tactical' | 'financial' | 'technical';
export type DecisionStatus = 'active' | 'archived' | 'implemented' | 'abandoned';
export type FactorCategory = 'market' | 'financial' | 'technical' | 'organizational' | 'customer' | 'competitive' | 'regulatory' | 'strategic';
export type ImpactLevel = 'high' | 'medium' | 'low';
export type Timeframe = 'short-term' | 'medium-term' | 'long-term';

// Factor Response
export interface Factor {
  id: string;
  decision_id: string;
  name: string;
  category: FactorCategory;
  impact_score: number;
  explanation?: string;
  created_at: string;
}

// Outcome Response
export interface Outcome {
  id: string;
  decision_id: string;
  description: string;
  outcome_type: string;
  likelihood?: number;
  impact?: ImpactLevel;
  timeframe?: Timeframe;
  status: string;
  created_at: string;
  updated_at: string;
}

// Decision Response
export interface Decision {
  id: string;
  user_id: string;
  title: string;
  description: string;
  category?: DecisionCategory;
  status: DecisionStatus;
  context?: Record<string, unknown>;
  graph_node_id?: string;
  created_at: string;
  updated_at: string;
  decision_date?: string;
  factors: Factor[];
  outcomes: Outcome[];
}

// Create Decision Request
export interface DecisionCreate {
  title: string;
  description: string;
  category?: DecisionCategory;
  decision_date?: string;
  context?: Record<string, unknown>;
}

// Update Decision Request
export interface DecisionUpdate {
  title?: string;
  description?: string;
  category?: DecisionCategory;
  status?: DecisionStatus;
  decision_date?: string;
  context?: Record<string, unknown>;
}

// Query Request
export interface DecisionQuery {
  query: string;
  include_timeline?: boolean;
  include_factors?: boolean;
  max_results?: number;
}

// Timeline Event
export interface TimelineEventAPI {
  date: string;
  type: string;
  title: string;
  is_main: boolean;
  relationship?: string;
}

// Timeline Response
export interface TimelineResponse {
  decision_id: string;
  decision_title: string;
  events: TimelineEventAPI[];
}

// Related Decision
export interface RelatedDecision {
  id: string;
  title: string;
  date: string;
  relationships: string[];
  distance: number;
}

// Related Decisions Response
export interface RelatedDecisionsResponse {
  decision_id: string;
  related: RelatedDecision[];
}

// Factor Analysis
export interface FactorAnalysis {
  name: string;
  category: FactorCategory;
  impact_score: number;
  explanation: string;
}

// Factor Analysis Response
export interface FactorAnalysisResponse {
  decision_id: string;
  factors: FactorAnalysis[];
}

// Predicted Outcome
export interface PredictedOutcome {
  description: string;
  likelihood: number;
  impact: ImpactLevel;
  timeframe: Timeframe;
}

// Outcome Prediction Response
export interface OutcomePrediction {
  decision_id: string;
  outcomes: PredictedOutcome[];
  risks: string[];
  opportunities: string[];
}

// Graph Stats
export interface GraphStats {
  decisions: number;
  people: number;
  projects: number;
  factors: number;
  relationships: number;
}

// Decision List Response
export interface DecisionList {
  decisions: Decision[];
  total: number;
  page: number;
  page_size: number;
}

// Entity Extraction
export interface EntityExtraction {
  decisions: string[];
  people: string[];
  projects: string[];
  dates: string[];
  keywords: string[];
}

// Analysis Response
export interface AnalysisResponse {
  decision: Decision;
  timeline?: TimelineResponse;
  factors?: FactorAnalysis[];
  predictions?: OutcomePrediction;
  related_decisions?: RelatedDecision[];
}

// Pattern Analysis (for Phase 6)
export interface DecisionPattern {
  pattern_type: string;
  description: string;
  frequency: number;
  decisions: string[];
  timespan: {
    start: string;
    end: string;
  };
}

export interface PatternAnalysisResponse {
  patterns: DecisionPattern[];
  insights: string[];
  recommendations: string[];
}

// AI Insights (for Phase 6)
export interface AIInsight {
  type: 'observation' | 'recommendation' | 'risk' | 'opportunity';
  title: string;
  description: string;
  confidence: number;
  related_factors?: string[];
}

export interface AIInsightsResponse {
  decision_id: string;
  insights: AIInsight[];
  summary: string;
  generated_at: string;
}
